src/AbstractCompiler.php

```diff
@@ -0,0 +1,672 @@
+                $this->getStringExpression($argumentB, $expressionB)
+            )
+        ) {
+            $expression = new OperatorLess($expressionA, $expressionB);
+            return true;
+        }
+
+        return false;
+    }
+
+    protected function getLessEqualFunction($argumentA, $argumentB, &$expression)
+    {
+        if (
+            (
+                $this->getNumericExpression($argumentA, $expressionA, $typeA) &&
+                $this->getNumericExpression($argumentB, $expressionB, $typeB)
+            ) || (
+                $this->getStringExpression($argumentA, $expressionA) &&
+                $this->getStringExpression($argumentB, $expressionB)
+            )
+        ) {
+            $expression = new OperatorLessEqual($expressionA, $expressionB);
+            return true;
+        }
+
+        return false;
+    }
+
+    protected function getGreaterFunction($argumentA, $argumentB, &$expression)
+    {
+        if (
+            (
+                $this->getNumericExpression($argumentA, $expressionA, $typeA) &&
+                $this->getNumericExpression($argumentB, $expressionB, $typeB)
+            ) || (
+                $this->getStringExpression($argumentA, $expressionA) &&
+                $this->getStringExpression($argumentB, $expressionB)
+            )
+        ) {
+            $expression = new OperatorGreater($expressionA, $expressionB);
+            return true;
+        }
+
+        return false;
+    }
+
+    protected function getGreaterEqualFunction($argumentA, $argumentB, &$expression)
+    {
+        if (
+            (
+                $this->getNumericExpression($argumentA, $expressionA, $typeA) &&
+                $this->getNumericExpression($argumentB, $expressionB, $typeB)
+            ) || (
+                $this->getStringExpression($argumentA, $expressionA) &&
+                $this->getStringExpression($argumentB, $expressionB)
+            )
+        ) {
+            $expression = new OperatorGreaterEqual($expressionA, $expressionB);
+            return true;
+        }
+
+        return false;
+    }
+
+    protected function getNumericBinaryFunction($name, $argumentA, $argumentB, &$expression, &$type)
+    {
+        if (
+            !$this->getNumericExpression($argumentA, $expressionA, $typeA) ||
+            !$this->getNumericExpression($argumentB, $expressionB, $typeB)
+        ) {
+            return false;
+        }
+
+        $aIsAnInteger = $typeA === Output::TYPE_INTEGER;
+        $bIsAnInteger = $typeB === Output::TYPE_INTEGER;
+        $aIsAFloat = $typeA === Output::TYPE_FLOAT;
+        $bIsAFloat = $typeB === Output::TYPE_FLOAT;
+
+        if ($name === 'plus' || $name === 'minus' || $name === 'times' || $name === 'divides') {
+            if ($aIsAnInteger && $bIsAnInteger) {
+                $type = Output::TYPE_INTEGER;
+            } elseif ($aIsAnInteger && $bIsAFloat || $aIsAFloat && $bIsAnInteger || $aIsAFloat && $bIsAFloat) {
```

Comment:
Make the operator of evaluation explicit. (And check for other instances of <operator> <operand> <operator>)